### ISSUE TITLE
Extend process API to allow multiple spans in a process

### ISSUE THREAD

**Issue by fholzhauser** [Bluehouse-Technology/otter#14]
Follow up from #12 
The current process API only allows handling 1 span in a process. There is however a valid use case to organise the collection of the request information in different spans in the process e.g. based on a the part of the code (module) the request handling passes. The proposed enhancement would add new functions for start/log/tag/id/get/finish that take the span name as first argument and store/manipulate the data in the process dictionary with a composite (tuple) key i.e. {otter_span_information, Name}.

**Comment by cmullaparthi** [comment]
Why not use the SpanId itself as the additional element instead of requiring a SpanName? And this is exactly the current state of the Functional API - granted you probably can't call it functional any more in the pure sense as it is using the process dictionary to accumulate span data.

The additional problem with using the process dictionary is that this span cannot be used across processes to accumulate data. Using ETS will be expensive. We could mitigate the impact of that by using an ETS bag table perhaps so that we insert one entry per API invocation, and only do a lookup when the span is 'finished'. Not as cheap as the pdict, but more versatile.

**Comment by fholzhauser** [comment]
I think the current separation of functional API and process dict API is quite clear and simple with the corresponding limitations/benefits (including the process boundary). They have their specific use cases that (I hope) are made clear in the descriptions, and most Erlangers can relate to. I don't think we should make it any more "fuzzy". If there is another API with some storage to come, I'd like to have it in addition (hence #13), not to lose the benefits of using the process dictionary.

Regarding the solution for allowing multiple spans in the process dict :  The motivation of the process dict API is to  minimise the code instrumentation impact by not having to pass around data while paying only minimal performance penalty. It also fits perfectly with the normal Erlang pattern (for such applications) of having request handling processes. The request and the process have the same life span (pun intended). Initially the idea was to represent a request process in a system with a single span with logs/tags to indicate timings (logs) and data collection (tags) for each relevant code piece the request hits in that process. 
I do understand however the potential "visual" benefit of having these parts of code indicated in different spans. The reason for proposing the SpanName is that it is expected to indicate the context of a particular span and it would normally be static in the code. It does not need to be "remembered" and passed in a variable. It would be the closest thing to keep the current fairly low instrumentation impact and allow having those different spans in the process for indicating different pieces of code.

**Comment by cmullaparthi** [comment]
I think we can all agree that the functional API is slightly cumbersome to use. While this is a contrived example, it illustrates the problem quite nicely.

```erlang
    S1 = otter:start("test_span"),
    S2 = otter:log(S1, "started"),
    S3 = otter:tag(S2, "result", "ok"),
    S4 = otter:log(S3, "α =:= ω"),
    S5 = otter:log(S4, 123456),
    S6 = otter:log(S5, 'this is a atom'),
    S7 = otter:log(S6, io_lib:format("int: ~w, float: ~f, hex: ~.16B, Span: ~p",
					  [1, 1.0, 1, S6])),
    S8 = otter:log(S7, S7),
    S9 = otter:log(S8, fun() -> "result of function" end),
```

Storing the span data within the pdict, changes this to
```erlang
    ok = otter:log(Span_id, "started"),
    ok = otter:tag(Span_id, "result", "ok"),
    ok = otter:log(Span_id, "α =:= ω"),
    ok = otter:log(Span_id, 123456),
    ok = otter:log(Span_id, 'this is a atom'),
    ok = otter:log(Span_id, io_lib:format("int: ~w, float: ~f, hex: ~.16B, Span: ~p",
					  [1, 1.0, 1, Span_id])),
    ok = otter:log(Span_id, Span_id),
    ok = otter:log(Span_id, fun() -> "result of function" end),
    ok = otter:finish(Span_id),
```

But this then gives rise to the limitation that the span is limited to the process in question. Would tihs be better solved by an additional function to retrieve the span so that it can be used by another process? I can't see a nice way of solving it. The proposed redesign, gives you a way of swapping the F-API with the P-API by just a simple search/replace whereas the current version in the master branch requires a more invasive change.

@RoadRunnr anything to add to this?

**Comment by fholzhauser** [comment]
### Short version
I think it is perfectly fine to add more API's to otter as long as we keep the current ones. The only way I can think of for improving/replacing the process API which keeps exactly the same benefits as the current one is to use the static span names (see below). 

### Long version
I'm not arguing that the functional API is unfriendly (as most pure functional APIs in Erlang). It is however the basic wrapper around the span data structure, so definitely here to stay. We may chose to not communicate it, but that does not make it go away. Keeping it exposed as an API however usually pleases purists ;-) and (as Heinz also mentioned) encourages writing wrappers around it. Actually under the hood the process dict API is also a wrapper. I'm not against adding more of these wrappers with a persistence of choice to otter (which is why #13 was created).

Regarding the example code with the Span_id API. I agree that it is definitely less ugly than the functional API. However I'm not quite sure that it is the most common scenario to have the whole request handling process implemented in 1 single function, not calling anything else to have the Span_id at hand all the time. Normally we'd need to add the extra Span_id argument (or more of them) to the functions we call and want to have log/tag added to the span in those. In that sense it is more impact on the code than the current process dict API (or the proposed change here to use the static name for different spans in the process). So the idea to use the Span_id with the process dictionary is a kind of hybrid solution. Which is probably also fine as another extra API. 
It don't believe however that it should replace the current process dictionary API. I think if we are to extend the API, then the static SpanName tag is still the one that provides the same instrumentation benefit while allowing multiple spans in the process. 
Regarding the possibility to switch from the pdict API to the functional, there is an api call to get the current span from the process dictionary. Probably it wouldn't be a lot of trouble to also add a function to push a span to the process dictionary either.

I'd propose that I'll work in the improvement described in this enhancement and we can continue discussing the Span_id API in #13.

**Comment by cmullaparthi** [comment]
How about we change the API so that it accepts either a span_id() or #span{} ? That way, we get the best of both worlds? So the type specs in otter.erl would look something as follows.

```erlang
-spec tag(span_id() | span(), info(), info()) -> ok.
-spec tag(span_id() | span(), info(), info(), service()) -> ok.
-spec log(span_id() | span(), info()) -> ok.
-spec log(span_id() | span(), info(), service()) -> ok.
-spec finish(span_id() | span()) -> ok.
-spec ids(span_id() | span()) -> {trace_id(), span_id()}.
```

@RoadRunnr @fholzhauser what do you think?

**Comment by cmullaparthi** [comment]
Actually the above proposal will suffer if multiple processes try to update the same span with some tag/log information. So sounds like we cannot get away from having a shared storage for spans. Even if we went down the functional route, it still doesn't solve the problem of two different processes adding data to a span concurrently.

**Comment by fholzhauser** [comment]
Chandru, let's move to #13 with the SpanId API discussion. As you also concluded we can't really do that consistently without some sort of storage. I'd like to keep the scope of this one to the extension of the process dict API with span name tagging.

**Comment by RoadRunnr** [comment]
----- On Apr 26, 2017, at 9:50 AM, Chandrashekhar Mullaparthi <notifications@github.com> wrote: 

> Actually the above proposal will suffer if multiple processes try to update the
> same span with some tag/log information. So sounds like we cannot get away from
> having a shared storage for spans. Even if we went down the functional route,
> it still doesn't solve the problem of two different processes adding data to a
> span concurrently.

Updating a span does not necessarily mean that you have to update a data structure. If your storage is ets, you could simply add a new entry for every log and tag event, and then collect all events that belong to a given span only when sending. Most operations will then be appending to the ets table, with only one operation (the final send), retrieving and deleting.

**Comment by fholzhauser** [comment]
Multiple span (with different names) process dictionary API is implemented in 0.3.0
